torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -49,21 +49,36 @@ kir::Bool* UnrollPass::getThreadPredicate(const kir::TensorView* tv) {
 }
 
 void UnrollPass::handle(kir::Expr* expr) {
-  // If tv op, predicate it (except for top level expressions)
-  if (ir_utils::isTVOp(expr) && !for_loops_.empty()) {
+  if (ir_utils::isTVOp(expr)) {
+    // If tv op, predicate it (except for top level expressions)
     const auto out_tv = expr->outputs()[0]->as<kir::TensorView>();
+    const bool should_predicate = !for_loops_.empty() ||
+        out_tv->memoryType() == MemoryType::Global ||
+        out_tv->memoryType() == MemoryType::Shared;
```

Comment:
That should be right.

Thinking about this again, wouldn't it make more sense to remove `!for_loops_.empty()`? I.e.,
```
const bool should_predicate = out_tv->memoryType() == MemoryType::Global ||
    out_tv->memoryType() == MemoryType::Shared;
```

Having thread predicates should be always safe. We can just omit them for local memory, whether expressions are inside loops or not, as it won't affect correctness.

Reply:
This is great point, thanks! 

I actually wanted to understand more before committing a change. I understand that if output is local then we don't care about thread in terms of correctness but not sure if that's the whole story, mainly because I'm still brushing up on GPUs. 

Say for example, the op reads global memory, would this predicate have perf implications, or are we going to generate code using similar pattern?

Reply:
Yeah, that could have some perf overheads. Not sure how much difference we would see in terms of performance. I think either is fine for now.

Reply:
We'll postpone this improvement to the next few PRs that'd revisit this part. Thanks for the great suggestions! 👍